snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java

```diff
@@ -0,0 +1,220 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.snowflake.entities.SnowflakeSchema;
+import org.apache.iceberg.snowflake.entities.SnowflakeTable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SnowflakeCatalog extends BaseMetastoreCatalog
+    implements Closeable, SupportsNamespaces, Configurable<Object> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SnowflakeCatalog.class);
+
+  private Object conf;
+  private String catalogName = SnowflakeResources.DEFAULT_CATALOG_NAME;
+  private Map<String, String> catalogProperties = null;
+  private FileIO fileIO;
+  private SnowflakeClient snowflakeClient;
+
+  public SnowflakeCatalog() {}
+
+  @VisibleForTesting
+  void setSnowflakeClient(SnowflakeClient snowflakeClient) {
+    this.snowflakeClient = snowflakeClient;
+  }
+
+  @VisibleForTesting
+  void setFileIO(FileIO fileIO) {
+    this.fileIO = fileIO;
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    LOG.debug("listTables with namespace: {}", namespace);
+    Preconditions.checkArgument(
+        namespace.length() <= SnowflakeResources.MAX_NAMESPACE_DEPTH,
+        "Snowflake doesn't support more than %s levels of namespace, got %s",
+        SnowflakeResources.MAX_NAMESPACE_DEPTH,
+        namespace);
+
+    List<SnowflakeTable> sfTables = snowflakeClient.listIcebergTables(namespace);
+
+    return sfTables.stream()
+        .map(
+            table ->
+                TableIdentifier.of(table.getDatabase(), table.getSchemaName(), table.getName()))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    throw new UnsupportedOperationException(
+        String.format("dropTable not supported; attempted for table '%s'", identifier));
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    throw new UnsupportedOperationException(
+        String.format("renameTable not supported; attempted from '%s' to '%s'", from, to));
+  }
+
+  @Override
+  public void initialize(String name, Map<String, String> properties) {
+    catalogProperties = properties;
+
+    if (name != null) {
+      this.catalogName = name;
+    }
+
+    if (snowflakeClient == null) {
+      String uri = properties.get(CatalogProperties.URI);
+      Preconditions.checkNotNull(uri, "JDBC connection URI is required");
+
+      try {
+        // We'll ensure the expected JDBC driver implementation class is initialized through
+        // reflection
+        // regardless of which classloader ends up using this JdbcSnowflakeClient, but we'll only
+        // warn if the expected driver fails to load, since users may use repackaged or custom
+        // JDBC drivers for Snowflake communcation.
+        Class.forName(JdbcSnowflakeClient.EXPECTED_JDBC_IMPL);
```

Comment:
Good question - hadn't had a chance to test it yet, but I've tested it successfully now on EMR, as well as the error case.

    # Works
    spark-shell --packages net.snowflake:snowflake-jdbc:3.13.22 --jars iceberg-spark-runtime-3.2_2.12-instance-initialization-forname.jar --conf ...

    # Error
    spark-shell --packages net.snowflake:snowflake-jdbc:3.13.22 --jars iceberg-spark-runtime-3.2_2.12-no-forname.jar --conf ...

I also tried with it in the static block like yours and it happened to work on my EMR cluster. However, as far as I understand it, the reason it's a gamble to put a `Class.forName` call in a static initalization block especially when using Spark is that there's a lot of fancy dynamic jar-downloading and classloading stuff, such that I wouldn't count on the classpath resources necessarily being complete at the time the static initializer block is called.

One possibility is if you tested with the Snowflake JDBC Driver being a `--jars` argument instead of `--packages` argument since I tested it in `--packages`?